src/frontend/context/SecurityContext.tsx

```diff
@@ -0,0 +1,159 @@
+import * as React from "react";
+import AsyncStorage from "@react-native-async-storage/async-storage";
+
+import { KILL_KEY, PASSWORD_KEY } from "../constants";
+import {
+  NavigationActions,
+  NavigationContainerComponent,
+} from "react-navigation";
+
+type AppModeTypes = "normal" | "kill";
+type AuthStatusTypes = "pending" | "authenticated" | "notRequired";
+
+export type AuthState = {
+  passcode: string | null;
+  appMode: AppModeTypes;
+  killModeEnabled: boolean;
+  authStatus: AuthStatusTypes;
+};
+
+type AuthActions =
+  | { type: "setPasscode"; newPasscode: string | null }
+  | { type: "toggleAppMode"; newAppMode?: AppModeTypes }
+  | { type: "toggleKillModeEnabled"; newKillModeValue?: boolean }
+  | { type: "setAuthStatus"; newAuthStatus: AuthStatusTypes };
+
+function securityReducer(state: AuthState, action: AuthActions): AuthState {
+  switch (action.type) {
+    /***
+     * if passcode is set to null we need to set the auth status to 'notRequired'
+     *
+     * If a passcode is set, we want to set the authStatus to 'authenticated'
+     * If the auth status is set to 'pending' they will be navigated to the auth screen, to type in their password.
+     */
+    case "setPasscode":
+      const passcode = action.newPasscode;
+      if (!validPasscode(passcode)) throw new Error("Invalid New Password");
+
+      AsyncStorage.setItem(PASSWORD_KEY, passcode || "");
+      if (!passcode) {
+        AsyncStorage.setItem(KILL_KEY, JSON.stringify(false));
+        return {
+          ...state,
+          passcode,
+          authStatus: "notRequired",
+          appMode: "normal",
+        };
```

Comment:
wondering if this should also set `killModeEnabled` to `false` i.e. turning off the app passcode also resets the kill mode state. right now, the following occurs:

1. Turn on passcode
2. Turn on kill mode
3. Turn off passcode
4. Turn on passcode
5. `killModeEnabled` is `true`

in 5, i'd kind of expect that to be `false` i.e. the app passcode determines how that state resets. thoughts?

Reply:
You are correct, good catch.